File: ios/Sources/BackgroundGeolocationPlugin/PrioritySyncManager.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 gachlab

import Foundation
import Network
import BackgroundGeolocationCore

/// Immediately POSTs safety-critical events to a priority endpoint, with
/// retry and offline queuing via NWPathMonitor.
final class PrioritySyncManager {

    // MARK: – Constants

    static let defaultRetries = 3
    static let defaultRetryDelays: [TimeInterval] = [10, 30, 60]
    static let defaultEvents = ["possibleCrash", "sos"]

    // MARK: – Config

    private let url: URL?
    private let headers: [String: String]?
    private let maxRetries: Int
    private let retryDelays: [TimeInterval]

    // MARK: – State

    private let queue = DispatchQueue(label: "gachlab.priority-sync", qos: .utility)
    private var sentTimestamps = Set<Int64>()
    private var sentOrder = [Int64]()          // insertion order so overflow drops the OLDEST
    private var offlineQueue: [(String, [String: Any])] = []
    private var monitor: NWPathMonitor?
    // Cached connectivity from the persistent monitor — avoids a per-submit probe race.
    private var isNetworkAvailable = true

    // MARK: – Init

    init(config: BGConfig) {
        let urlString = config.prioritySyncUrl?.isEmpty == false
            ? config.prioritySyncUrl!
            : (config.url ?? "")
        self.url = URL(string: urlString)
        self.headers = config.httpHeaders
        self.maxRetries = config.prioritySyncRetries ?? Self.defaultRetries
        self.retryDelays = config.prioritySyncRetryDelays?.map { TimeInterval($0) }
            ?? Self.defaultRetryDelays
        if self.url != nil { startMonitoring() }
    }

    deinit { monitor?.cancel() }

    // MARK: – Public API

    /// Submit an event for immediate delivery. No-op when no URL is configured
    /// or when the event timestamp was already submitted (dedup).
    func submit(eventType: String, payload: [String: Any]) {
        guard url != nil else { return }
        // Accept any numeric timestamp (Int, Double, NSNumber bridged from JS), matching
        // Android's optLong — a strict `as? Int64` would miss most bridged numbers and break dedup.
        let ts = (payload["timestamp"] as? NSNumber)?.int64Value
            ?? Int64(Date().timeIntervalSince1970 * 1000)
        queue.async { [weak self] in
            guard let self else { return }
            guard self.sentTimestamps.insert(ts).inserted else { return }
            self.sentOrder.append(ts)
            if self.sentTimestamps.count > 200 {
                // Drop the oldest 100 deterministically (mirrors Android's LinkedHashSet trim).
                let drop = self.sentOrder.prefix(100)
                self.sentOrder.removeFirst(min(100, self.sentOrder.count))
                for old in drop { self.sentTimestamps.remove(old) }
            }
            if !self.isConnected() {
                self.offlineQueue.append((eventType, payload))
                return
            }
            self.postWithRetry(eventType: eventType, payload: payload, attempt: 1)
        }
    }

    // MARK: – Private

    private func postWithRetry(eventType: String, payload: [String: Any], attempt: Int) {
        guard let url else { return }
        let code = httpPost(url: url, headers: headers, body: payload)
        if (200...299).contains(code) {
            NotificationCenter.default.post(
                name: .BGPrioritySyncSuccess,
                object: nil,
                userInfo: ["eventType": eventType, "attemptNumber": attempt]
            )
        } else if attempt < maxRetries {
            let delay = retryDelays.indices.contains(attempt - 1)
                ? retryDelays[attempt - 1]
                : (retryDelays.last ?? 60)
            queue.asyncAfter(deadline: .now() + delay) { [weak self] in
                self?.postWithRetry(eventType: eventType, payload: payload, attempt: attempt + 1)
            }
        } else {
            NotificationCenter.default.post(
                name: .BGPrioritySyncFailed,
                object: nil,
                userInfo: ["eventType": eventType, "httpStatus": code, "attempts": attempt]
            )
        }
    }

    // Reads the cached state kept current by the persistent monitor. Called on `queue`,
    // where the flag is also written, so no probe and no race.
    private func isConnected() -> Bool { isNetworkAvailable }

    private func startMonitoring() {
        let m = NWPathMonitor()
        m.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            let available = path.status == .satisfied
            self.queue.async {
                self.isNetworkAvailable = available
                if available { self.flushOfflineQueue() }
            }
        }
        m.start(queue: DispatchQueue(label: "gachlab.psm.monitor"))
        monitor = m
    }

    private func flushOfflineQueue() {
        guard !offlineQueue.isEmpty else { return }
        let pending = offlineQueue
        offlineQueue.removeAll()
        pending.forEach { (type, payload) in
            postWithRetry(eventType: type, payload: payload, attempt: 1)
        }
    }

    private func httpPost(url: URL, headers: [String: String]?, body: [String: Any]) -> Int {
        guard let data = try? JSONSerialization.data(withJSONObject: body) else { return -1 }
        var request = URLRequest(url: url, timeoutInterval: 30)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        headers?.forEach { request.setValue($1, forHTTPHeaderField: $0) }
        request.httpBody = data

        var responseCode = -1
        let semaphore = DispatchSemaphore(value: 0)
        URLSession.shared.dataTask(with: request) { _, response, _ in
            responseCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            semaphore.signal()
        }.resume()
        _ = semaphore.wait(timeout: .now() + 35)
        return responseCode
    }
}

File: android/src/main/java/com/gachlab/geolocation/network/PrioritySyncManager.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 gachlab

package com.gachlab.geolocation.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.gachlab.geolocation.BGConfig
import com.gachlab.geolocation.ServiceEvent
import org.json.JSONObject
import java.util.Collections
import java.util.LinkedHashSet
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executors

/**
 * Immediately POSTs safety-critical events to a priority endpoint, with exponential
 * retry and offline queuing via ConnectivityManager.NetworkCallback.
 *
 * Each [submit] call is fire-and-forget; results arrive via [onEvent].
 */
internal class PrioritySyncManager(
    context: Context,
    config: BGConfig,
    private val onEvent: (ServiceEvent) -> Unit,
) {
    private val appContext = context.applicationContext
    private val url: String = config.prioritySyncUrl?.takeIf { it.isNotEmpty() }
        ?: config.url?.takeIf { it.isNotEmpty() } ?: ""
    private val headers: Map<String, String>? = config.httpHeaders
    private val maxRetries: Int = config.prioritySyncRetries ?: DEFAULT_RETRIES
    private val retryDelays: List<Long> = config.prioritySyncRetryDelays ?: DEFAULT_RETRY_DELAYS

    private val executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())
    // LinkedHashSet preserves insertion order so we can trim oldest entries on overflow.
    private val sentTimestamps: MutableSet<Long> = Collections.synchronizedSet(LinkedHashSet())
    private val offlineQueue = CopyOnWriteArrayList<Pair<String, JSONObject>>()

    private val connectivityManager =
        appContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    init { registerConnectivityCallback() }

    /**
     * Submit an event for priority delivery. No-op when [url] is empty or when the
     * event timestamp has already been submitted (dedup).
     */
    fun submit(eventType: String, payload: JSONObject) {
        if (url.isEmpty()) return
        val ts = payload.optLong("timestamp", System.currentTimeMillis())
        synchronized(sentTimestamps) {
            if (!sentTimestamps.add(ts)) return
            if (sentTimestamps.size > MAX_DEDUP_SIZE) {
                val iter = sentTimestamps.iterator()
                repeat(MAX_DEDUP_SIZE / 2) { if (iter.hasNext()) { iter.next(); iter.remove() } }
            }
        }
        if (!isConnected()) {
            offlineQueue.add(Pair(eventType, payload))
            Log.d(TAG, "Offline — queued priority event: $eventType (${offlineQueue.size} pending)")
            return
        }
        executor.submit { postWithRetry(eventType, payload, 1) }
    }

    fun destroy() {
        networkCallback?.let {
            try { connectivityManager.unregisterNetworkCallback(it) } catch (_: Exception) {}
        }
        executor.shutdown()
    }

    private fun postWithRetry(eventType: String, payload: JSONObject, attempt: Int) {
        val client = HttpClient(url, headers = headers)
        val code = client.postJSON(payload)
        Log.d(TAG, "PrioritySync $eventType attempt $attempt → HTTP $code")
        when {
            code in 200..299 ->
                onEvent(ServiceEvent.PrioritySyncSuccess(eventType, attempt))
            attempt < maxRetries -> {
                val delay = retryDelays.getOrElse(attempt - 1) { retryDelays.lastOrNull() ?: 60_000L }
                mainHandler.postDelayed(
                    { executor.submit { postWithRetry(eventType, payload, attempt + 1) } },
                    delay
                )
            }
            else ->
                onEvent(ServiceEvent.PrioritySyncFailed(eventType, code, attempt))
        }
    }

    private fun isConnected(): Boolean =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val net = connectivityManager.activeNetwork ?: return false
            connectivityManager.getNetworkCapabilities(net)
                ?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
        } else {
            @Suppress("DEPRECATION")
            connectivityManager.activeNetworkInfo?.isConnected == true
        }

    private fun registerConnectivityCallback() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) return
        networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                val pending = offlineQueue.toList()
                if (pending.isEmpty()) return
                offlineQueue.clear()
                Log.d(TAG, "Connectivity restored — flushing ${pending.size} queued priority events")
                pending.forEach { (type, payload) ->
                    executor.submit { postWithRetry(type, payload, 1) }
                }
            }
        }.also { cb ->
            try {
                val req = NetworkRequest.Builder()
                    .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                    .build()
                connectivityManager.registerNetworkCallback(req, cb)
            } catch (e: Exception) {
                Log.w(TAG, "Could not register network callback: ${e.message}")
            }
        }
    }

    companion object {
        private const val TAG = "PrioritySyncManager"
        private const val MAX_DEDUP_SIZE = 200
        const val DEFAULT_RETRIES = 3
        val DEFAULT_RETRY_DELAYS = listOf(10_000L, 30_000L, 60_000L)
        val DEFAULT_EVENTS = listOf("possibleCrash", "sos")
    }
}
